File: LeetCode_Record/2299.强密码检验器II.cpp
#include <string>
using namespace std;

class Solution {
public:
    bool strongPasswordCheckerII(string password) {
        if(password.size() < 8)
            return false;
        bool upper = false;
        bool lower = false;
        bool digit = false;
        bool sign = false;
        char prev = ' ';
        for(auto& i:password){
            if(i == prev)
                return false;
            prev = i;
            if(isdigit(i))
                digit = true;
            else if(isupper(i))
                upper = true;
            else if(islower(i))
                lower = true;
            else
                sign = true;
        }
        if(upper && lower && digit && sign)
            return true;
        return false;
    }
};

File: LeetCode_Record/2299.强密码检验器II.java
class Solution {
    private boolean sign, upper, lower, digit;

    public boolean strongPasswordCheckerII(String password) {
        int n = password.length();
        if(n < 8)
            return false;
        Character prev = password.charAt(0);
        charIs(prev);
        for(int i = 1; i < n; i++){
            Character now = password.charAt(i);
            if(now == prev)
                return false;
            prev = now;
            charIs(now);
        }
        if(sign && upper && lower && digit)
            return true;
        return false;
    }

    public void charIs(Character c){
        if(Character.isUpperCase(c))
            upper = true;
        else if(Character.isLowerCase(c))
            lower = true;
        else if(Character.isDigit(c))
            digit = true;
        else 
            sign = true;
    }
}